core/matrix/diagonal.cpp

```diff
@@ -302,4 +318,24 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_DIAGONAL_MATRIX);
 
 
 }  // namespace matrix
+
+
+// Implement DiagonalExtractable for LinOp when Diagonal is complete class
+template <typename ValueType>
+std::unique_ptr<LinOp> DiagonalExtractable<ValueType>::extract_diagonal_linop()
+    const
+{
+    auto diag = this->extract_diagonal();
+    auto p = dynamic_cast<LinOp *>(diag.get());
+    diag.release();
+    return std::unique_ptr<LinOp>{p};
```

Comment:
Isn't it possible to return `diag` directly? std::unique_ptr has a up-casting constructor that should take care of that.